✨ Implements Backup and Restore for objectMover

jpmcb opened the pull request:
### What this PR does / why we need it:

Implements the Backup and Restore methods on the objectMover

Backup will iterate objects in a given namespace, building the object graph, to then save them to a provided directory as a set of json files. These files will appear as:
```
{configured-directory}/object-kind_object-name_object-namespace.yaml
```
Restore will iterate object files using a given glob in a configured directory to re-build an object graph. Once all files are iterated and the object graph is created, re-applies those objects to a given management / bootstrap cluster with provided client.

These methods can then be used to create a "backup / restore" cycle where a clusters objects are saved and restored latter.

Fixes #3441

This PR supersedes #4786 which was based on the 0.3 branching. Hope to get this into the 0.4.0 release. Since a previous build of my testing code / environment was based on 0.3, I wasn't able to test this out as fully as I'd like and would like guidance on testing this more. Definitely want feedback on this and open to any proposed changes!
cc: @fabriziopandini @vincepri

fabriziopandini commented:
/ok-to-test

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -592,6 +805,103 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
+	return nil
+}
+
+func (o *objectMover) restoreTargetObject(nodeToCreate *node, toProxy Proxy) error {
```
I suggest to change a little bit this func to make it possible to restart a restore after an error.
In order to do so we should try to get the object as a first thing; if it already exists, then it is required only to update the UID in order to make it possible o rebuild the owner chain.

fabriziopandini replied:
Let's have a unit test for this func ensuring:
- It creates objects as expected
- It does not overrides existing objects (but it fetches the the UID)

fabriziopandini replied:
Unit test are in place, but IMO we still need to improve the restoreTargetObject func.
My idea here here is that what is already in the cluster should not be touched by restore (because what is in the cluster is more recent that what is the files)
Accordingly, if get returns an object, we should only update the node inside the object graph with current object UID, so the dependants owner chain get properly recreated, but we should not touch the object itself in the cluster.
Does this seems reasonable to you?

jpmcb replied:
I see - I think that helps clarify. So instead of doing `Update` if the node already exists, we simply continue since we've updated the reference chain. I think all that would need to be changed is [this part here](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-3458bf9e9c245fe801573500c33de535f05d3a5887f080a405b819e3c0fed415R922-R925) since before this, we attempt a `Get` and update the objects UID based on any existing object. Then, we rebuild the owner references with the UID between L:883-903. We then attempt a `Create` by default and instead of it updating if the object already exists, simply continue (maybe log something out).

fabriziopandini replied:
I think you got the idea right. With regard to the implementation I struggle I little bit to follow what you are suggesting (the link above does not work for me 😞 ), but hopefully this can help.
- Get the object corresponding to the node to create
- If the object already exists, 
   - log a message
   - update `nodeToCreate.newUID=existingTargetObj.GetUID()` so the owner chain will be properly rebuilt 
  - return
- if the object does not exists, create it (in this case apierrors.IsAlreadyExists(err) should never happen, so you can cleanup a good chunk of the existing logic)

Note. This should be covered by a test

jpmcb replied:
Let me know what you think of my implementation for this! Happy to refactor :smile:

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -592,6 +805,103 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
+	obj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	obj.SetKind(nodeToCreate.identity.Kind)
+
+	// New objects cannot have a specified resource version. Clear it out.
+	obj.SetResourceVersion("")
+
+	// Removes current OwnerReferences
+	obj.SetOwnerReferences(nil)
+
+	// Rebuild all the OwnerReferences using the newUID of the owner nodes.
+	if len(nodeToCreate.owners) > 0 {
+		ownerRefs := []metav1.OwnerReference{}
+		for ownerNode := range nodeToCreate.owners {
+			ownerRef := metav1.OwnerReference{
+				APIVersion: ownerNode.identity.APIVersion,
+				Kind:       ownerNode.identity.Kind,
+				Name:       ownerNode.identity.Name,
+				UID:        ownerNode.newUID, // Use the owner's newUID read from the target management cluster (instead of the UID read during discovery).
+			}
+
+			// Restores the attributes of the OwnerReference.
+			if attributes, ok := nodeToCreate.owners[ownerNode]; ok {
+				ownerRef.Controller = attributes.Controller
+				ownerRef.BlockOwnerDeletion = attributes.BlockOwnerDeletion
+			}
+
+			ownerRefs = append(ownerRefs, ownerRef)
+		}
+
+		obj.SetOwnerReferences(ownerRefs)
+	}
```
This code seems a duplication of what we have in createTargetObject. What about creating a utility func and share?

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -102,6 +110,100 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
 	return o.move(objectGraph, proxy)
 }
 
+func (o *objectMover) Save(namespace string, directory string) error {
```
We should have a test for this func.
Given that we already have coverage for most of the utilities used internally I would suggest to test save with a fairly simple cluster, like e.g. https://github.com/kubernetes-sigs/cluster-api/blob/88b3ee3e1ac58440bc450126922a723cbc86f7ac/cmd/clusterctl/client/cluster/mover_test.go#L169-L179
checking that the expected files are created in a temp folder.

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -102,6 +110,100 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	return nil
+}
+
+func (o *objectMover) Restore(namespace string, glob string, directory string) error {
```
We should have a test for this func.
Given that this is the reverse too the save test, you can eventually combine the two tests, but I leave this to you

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -102,6 +110,100 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	}
+
+	// Gets all the types defined by the CRDs provided by file plus the ConfigMap/Secret core types.
+	err = objectGraph.getDiscoveryFileTypes(objs)
```
Might be I'm missing something, but why not use getDiscoveryTypes that reads from CRDs?

jpmcb replied:
`getDiscoveryTypes` uses a client to read the CRDs off of a management cluster. If we are restoring a management cluster from files, there would be no actual cluster to read CRDs from. Instead, `getDiscoveryFileTypes` _reads CRDs from objects passed in from files_. The intent was to do much of the same thing as `getDiscoveryTypes`, but instead, do it from a grouping of objects from files.

fabriziopandini replied:
According to the save/backup logic in this PR, the files contains only cluster object, machine deployment object etc. but not the actual Cluster API CRDs nor the Cluster API deployments/reach rules/ web hook configurations required to run the controllers.
As a consequence I assume that the something (e.g clusterctl) should install Cluster API (including CRDs) before doing restore on an empty cluster, otherwise restore will fail because Cluster API is not installed.
Am I wrong?

fabriziopandini replied:
Just to clarify, in restore CRD should exist in the target cluster, but they are there...

jpmcb replied:
Ah I see - that makes sense. So, if I'm understanding, the `getDiscoveryTypes` method should reach out to a target management cluster (be it in a move or a restore), and ensure that the Cluster API CRDs are available, present? 

I think your assumption is correct that in a restore, the CRDs _should_ already be there before a restore is attempted

fabriziopandini replied:
I think we should make getDiscoveryTypes flexible enough to reach out for the source or the target cluster, and use the latter only for restore. Another option a little bit less explicit is to enforce source=target for restore

fabriziopandini replied:
I did a little bit more of digging in this PR and in Restore and I L154 we are doing
```
objectGraph := newObjectGraph(o.fromProxy, o.fromProviderInventory)
```
also in L200 we are doing
```
if err := o.restore(objectGraph, o.fromProxy); err != nil {
```
That's makes me think that o.fromProxy in fact is the target proxy for restore, which is kind of confusing.
What about passing `toCluster Client`, explicitly to restore, like we are doing for move?

fabriziopandini commented on cmd/clusterctl/client/cluster/objectgraph.go:
```diff
@@ -293,6 +303,41 @@ func (o *objectGraph) getDiscoveryTypes() error {
 
+// getDiscoveryFileTypes returns the list of TypeMeta to be considered for a save / restore cycle.
+// This list includes all the types defined by a slice of objects populated from file
+func (o *objectGraph) getDiscoveryFileTypes(objs []unstructured.Unstructured) error {
```
I'm not sure that resuming type informations for object instances is the right approach, because we are missing all the information that comes from annotations at CRD level, which are getting more and more relevant in the move logic.
As commented above, let's try to use getDiscoveryTypes reading from CRD.

fabriziopandini commented on cmd/clusterctl/client/cluster/objectgraph.go:
```diff
@@ -78,6 +78,10 @@ type node struct {
+
+	// rawObject holds the object that is referenced when creating the node.
+	// the object can then be referenced latter in a save / restore cycle
+	rawObject *unstructured.Unstructured
```
Might be I'm missing something, but I don't think this field is necessary.
It appears that it is set during restore and the used only in getDiscoveryFileTypes, which is called before restore...

jpmcb replied:
The `rawObject` field is used in the restoreTargetObject method.

Since these objects are being restored from file, we need the original, raw object from the object file originating from the backed up management cluster. In my method, I set the object graph to have references to these original objects since we can't populate the object with the typical `get` method using a client (like we do in the move sequence).

Without this field, I'm not sure how we would accurately restore from file.

fabriziopandini replied:
Now that getDiscoveryFileTypes has been removed, should we remove this field?

jpmcb replied:
I believe this field is still essential. 

Without it, we would have no actual object reference when restoring objects. It is used [here](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-3458bf9e9c245fe801573500c33de535f05d3a5887f080a405b819e3c0fed415R864) as part of applying an object to a target cluster during restore. In the `move` method, we can generate some object key and query the source management cluster to fill in the rest of the object. During a restore, we cannot rely on a source management cluster to populate objects. So, without a source management cluster to interrogate, we have to rely on the raw object given from the file and populated when the object was added to the graph.

fabriziopandini replied:
Ok, got your point, however I think we should avoid storing the objects always when this is necessary only for restore.
What about:
- renaming this field to restoredObject
- introduce a addrestoredObj function that calls addObj and set restoredObject, and use this new func in restore
- remove proposed changes in `objToNode` and in `ownerToVirtualNode` so we avoid allocating objects in all the other operations

fabriziopandini replied:
This is now done

fabriziopandini reviewed (commented):
@jpmcb this is a very welcome feature and a great first PR in CluserAPI ! Thanks 🥳!

I left some comments, but the overall approach seems ok.

WRT to how to tests, I have left some feedback inline; Additionally, most probably this feature deserve an E2E test. However how this test should be implemented deserve some thinking, in order to not be destructive for other tests running in parallel. I will follow up on this point.

jpmcb commented on cmd/clusterctl/client/client.go:
```diff
@@ -51,6 +51,12 @@ type Client interface {
 	Move(options MoveOptions) error
 
+	// Save saves all the Cluster API objects existing in a namespace (or from all the namespaces if empty) to a target management cluster.
+	Save(options SaveOptions) error
```
Sounds good! Will go ahead with using `Backup`

jpmcb replied:
Sounds good! Now using `Backup`

jpmcb commented:
@fabriziopandini Thanks for all the great feedback.

I've added tests you recommended and added the various utility functions as requested. I think this should be ready for another round of reviewing 

The last changes I made include
- Use `backup` instead of `save` terminology
- Implement tests
- Reduce code duplication with a few utility funcs
- Added both `Backup` and `Restore` to the CLI as requested in the original issue
- Correctly pause before the backup and restore happens. Also unpausing after each
- Updated `restoreTargetObject` to be more resilient. Now looks for an existing object and updates the object being restored to the UID of the object it found. Then re-builds the owner references as needed for existing objects
- Removed `getDiscoveryFileTypes` in favor of `getDiscoveryTypes` since we want to interrogate the target cluster for CRD types, and not just the files that are being used in a restore cycle
- Now passes a `toCluster Client` to the restore method resolving some confusion from using `fromCluster`

Thanks so much for the time spent looking at this!

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +69,129 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	return o.restore(objectGraph, proxy)
+}
+
+func (o *objectMover) processFilesToObjs(glob string, dir string) ([]unstructured.Unstructured, error) {
```
what about 
```suggestion
func (o *objectMover) filesToObjs(glob string, dir string) ([]unstructured.Unstructured, error) {
```

jpmcb commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,391 @@ var moveTests = []struct {
+				fmt.Printf("Actual file content %v\n", string(fileContents))
+				g.Expect(string(fileContents)).To(Equal(expectedFileContents))
+
+				time.Sleep(time.Second)
```
I'm not sure. Without the sleep, I get this error:
```
--- FAIL: Test_objectMover_backupTargetObject (0.01s)
    --- FAIL: Test_objectMover_backupTargetObject/Cluster (0.01s)
        mover_test.go:587:
            Expected
                <time.Time>: 2021-06-21T14:55:32.36010393-06:00
            to be <
                <time.Time>: 2021-06-21T14:55:32.36010393-06:00

```
But with the sleep, it succeeds. I didn't dive it, but I added the sleep thinking that the tests and file operations may be happening so fast that there is no relative difference in time.

Interestingly, this fails:
```go
time.Sleep(time.Millisecond)
```
while this succeeds:
```go
time.Sleep(time.Millisecond * 5)
```

I believe the test is correct, but I'm not certain what golang or ginkgo internals are causing the time comparison to fail without a short sleep. I'm in favor of leaving it, or at least changing it to a short successful time period with a comment annotating what it's for.

fabriziopandini replied:
Ok, thanks for the explanation
Let's add a comment keeping track why we need this
```
// Add a delay so we ensure the file ModTime of the updated files is differente from the one in the original files
time.Sleep(time.Millisecond * 5)
```
Or something similar

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +341,75 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	moveSequence := getMoveSequence(graph)
+
+	// Create all objects group by group, ensuring all the ownerReferences are re-created.
+	for groupIndex := 0; groupIndex < len(moveSequence.groups); groupIndex++ {
```
Added a log line here indication the objects are being restore and also one in the `filesToObjs` func that is indicating what directory files are being restored from

jpmcb commented on cmd/clusterctl/cmd/restore.go:
```diff
@@ -0,0 +1,76 @@
+	},
+}
+
+func init() {
```
Good catch. Added the directory flag to restore. Now a user can specify the directory to load object files from. If they don't specify the directory, it'll use the default Cluster API config directory under the objects folder. This was [implemented in the setObjectsDir func](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-972d17a2075f3710e12506a9a27cb2b1a0f67b6b69dbc7eb342d1adff36a7d2eR197-R213)

jpmcb commented on cmd/clusterctl/cmd/backup.go:
```diff
@@ -0,0 +1,74 @@
+		"Context to be used within the kubeconfig file for the source management cluster. If empty, current context will be used.")
+	backupCmd.Flags().StringVarP(&buo.namespace, "namespace", "n", "",
+		"The namespace where the workload cluster is hosted. If unspecified, the current context's namespace is used.")
+	backupCmd.Flags().StringVar(&buo.directory, "directory", "",
```
I didn't make the directory flag required since the `setObjectsDir` func will default to the Cluster API default config folder under the `objects` directory if no directory is provided. This was implemented [here](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-972d17a2075f3710e12506a9a27cb2b1a0f67b6b69dbc7eb342d1adff36a7d2eR197-R213). If we expect the user to provide a directory no matter what, then maybe we remove this "default" logic from this func and cobra will error early if the flag is not provided

fabriziopandini replied:
I think removing the defaulting logic somehow makes sense, especially given that restore takes every yaml in a folder.
I'm +1 to ask the user to be explicit about the directory to use when doing backup/restore

jpmcb replied:
:+1: Removed glob and now require the `directory` flag for both backup and restore

jpmcb commented on cmd/clusterctl/cmd/restore.go:
```diff
@@ -0,0 +1,76 @@
+		"Context to be used within the kubeconfig file for the target management cluster. If empty, current context will be used.")
+	restoreCmd.Flags().StringVarP(&ro.namespace, "namespace", "n", "",
+		"The namespace to restore Cluster API objects to in the target management cluster. If unspecified, the current context's namespace is used.")
+	restoreCmd.Flags().StringVar(&ro.glob, "glob", "",
```
Good question. 

I think the general use case would be if a user wants to restore a certain cluster from a certain namespace. Let's look at a few examples:

 imagine I have `cluster-a` and `cluster-b` in namespace `my-namespace`. If I backup both clusters using the following command

```
clusterctl backup --namespace my-namespace
```

then both clusters in that namespace will be backed-up to the default config / objects directory. I end up with this kind of filetree (leaving out some of the other object files like `Secrets` for simplicity):

```
~/.cluster-api/objects
    | - Cluster_cluster-a_my-namespace.yaml
    | - Cluster_cluster-b_my-namespace.yaml
```

If I then want to restore _only_ `cluster-a`, I have no ability to do that unless I provide a glob. This would look like:

```
clusterctl restore --glob cluster-a
``` 

And it will only match files (and restore those objects) that match "cluster-a" as a substring.

Further, I can imagine the use case where only certain namespaces are restored. For example, if I have `cluster-foo` and `cluster-bar` in namespace `namespace-a` and `cluster-baz` in namespace `namespace-b`, I could backup both namespaces with:
```
clusterctl backup --namespace namespace-a
clusterclt backup --namespace namespace-b
```

Then, the default config objects directory would look like:

```
~/.cluster-api/objects
    | - Cluster_cluster-foo_namespace-a.yaml
    | - Cluster_cluster-bar_namespace-a.yaml
    | - Cluster_cluster-baz_namespace-b.yaml
```

Then, I could choose to restore _only_ those clusters that were in `namespace-a` with:
```
clusterctl restore --glob namespace-a
```
This could be _refined even further_ by providing both the namespace and the cluster name:
```
clusterctl restore --glob cluster-foo_namespace-a
```
will only restore objects that are associated with cluster `cluster-foo` from `namespace-a`.

This probably warrants being documented better in the help message. If this sound agreeable, I can go ahead and flesh out the help message

fabriziopandini replied:
Uhmm
I'm not sure the back/restore single file scenario really holds, given that the naming schema is kind-name-namespace, and so with the example above you are only to restore the cluster object, but not the entire hierarchy of objects belonging to a cluster.
WRT to restore a single namespace at time, this could make sense, but TBH I think we should reccomand to backup different namespaces in different folders.
Given that, my personal opinion is to remove this flag/parameter for the first iteration and keep the things simple, but it would be interesting to have some other opinions here.
@vincepri

jpmcb replied:
Good point - I think removing the glob option for now and requiring the directory would be a good first pass. And in the future, if we wanted some kind of partial restore or glob based restore, we can add it in.

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -31,13 +35,18 @@ import (
 type ObjectMover interface {
 	// Move moves all the Cluster API objects existing in a namespace (or from all the namespaces if empty) to a target management cluster.
 	Move(namespace string, toCluster Client, dryRun bool) error
+	// Backup saves all the Cluster API objects existing in a namespace (or from all the namespaces if empty) to a target management cluster.
```
Q: do we have tests for backup/restore all namespaces?

jpmcb replied:
Done - added a test for this

fabriziopandini commented:
@jpmcb thanks for bearing with me with the first rounds of reviews, really appreciated.
I think we are almost there, and IMO the only missing actions are:
- Simplifying UX removing partial restore.
- Make sure we store raw objects only for the restore operation.
- Make the restore logic to not override existing flags.

jpmcb commented:
@fabriziopandini no problem!!! Thank _you_ so much for your help and detailed reviews on this. You've made it easy to go back and refactor with pointed feedback. Let me know what else this needs and I'd be happy to modify as needed!

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +338,70 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	}
+
+	// Reset the pause field on the Cluster object in the target management cluster, so the controllers start reconciling it.
+	log.V(1).Info("Resuming the target cluster")
```
```suggestion
	log.V(1).Info("Resuming the source cluster")
```

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +338,70 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	}
+
+	// Reset the pause field on the Cluster object in the target management cluster, so the controllers start reconciling it.
+	log.V(1).Info("Resuming the source cluster")
```
```suggestion
	log.V(1).Info("Resuming the target cluster")
```
Also, should we Pause before restoring?

fabriziopandini commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -592,6 +772,132 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
+		return err
+	}
+
+	// Rebuild the source object
+	obj := nodeToCreate.restoreObject
+	obj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	obj.SetKind(nodeToCreate.identity.Kind)
+	objKey := client.ObjectKey{
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	// New objects cannot have a specified resource version. Clear it out.
+	obj.SetResourceVersion("")
+
+	// Removes current OwnerReferences
+	obj.SetOwnerReferences(nil)
```
What about moving this chunk of code after Get (to ~L860)?

jpmcb replied:
Sure thing - moved this below the `Get` so the two blocks are alittle more split up

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +68,127 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	}
+
+	for _, obj := range objs {
+		objectGraph.addObj(&obj)
```
Good catch!! :sweat_smile:

jpmcb commented on cmd/clusterctl/client/cluster/objectgraph.go:
```diff
@@ -133,13 +141,28 @@ func (o *objectGraph) addObj(obj *unstructured.Unstructured) {
+// addRestoredObj adds a Kubernetes object to the object graph that is generated during the restore discovery phase.
+// Populates the restoredObject field to be referenced during restore
+// During add, OwnerReferences are processed in order to create the dependency graph.
+func (o *objectGraph) addRestoredObj(obj *unstructured.Unstructured) {
```
Hmmm good point. I'm learning alot about the object graph here! I used your suggestion, but also kept the `processOwnerReferences` code chunk since I couldn't get tests to pass without it. Specifically, the "genericInfrastructure" object would be missing and I guessing that without processing the owner references, the cluster from file won't have the correct reference to objects it owns.

fabriziopandini replied:
this is now addressed

jpmcb commented on cmd/clusterctl/client/cluster/objectgraph_test.go:
```diff
@@ -1378,6 +1379,67 @@ func getObjectGraphWithObjs(objs []client.Object) *objectGraph {
 
+// Similar to the objectgraph Discvoery method, but populates objects using the restoreAddObj method
+// This then can be used in testing to accurately build an object graph as if from file
+func restoreDiscovery(o *objectGraph, namespace string) error {
```
Good call. This didn't make sense. I ended up removing this and added code in the tests to populate a temporary directory with the required object files (much like it would actually work) instead of relying on a bespoke `discovery` method that adds objects

jpmcb commented on cmd/clusterctl/client/move.go:
```diff
@@ -85,3 +115,83 @@ func (c *clusterctlClient) Move(options MoveOptions) error {
+	return nil
+}
+
+func (c *clusterctlClient) setObjectsDir(dir string) error {
```
Done! :heavy_check_mark:

jpmcb commented:
Hi @fabriziopandini - curious where we're at with this. Happy to continue to iterate and work out any other kinks you might see!

fabriziopandini reviewed (commented):
@jcmcb apart from a last question about restore on a not empty namespace below and a couple of nits, IMO this is ready to go

jpmcb commented on cmd/clusterctl/client/cluster/objectgraph.go:
```diff
@@ -132,14 +140,40 @@ func (o *objectGraph) addObj(obj *unstructured.Unstructured) {
+	node.restoreObject = obj.DeepCopy()
+
+	// Process OwnerReferences; if the owner object does not exists yet, create a virtual node as a placeholder for it.
+	o.processOwnerReferences(obj, node)
```
Hmm, `processOwnerReferences` isn't getting called inside the `objToNode` func. And I don't think calling `processOwnerReferences` will work inside the `objToNode` func since we may return early if an existing node is already found. I arbitraily tried to move `processOwnerReferences` to the end of `objToNode` but it had a number of failures in the object graph testing suite

fabriziopandini replied:
You are right. sorry I got confused

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +68,129 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	log := logf.Log
+	log.Info("Performing restore...")
+
+	objectGraph, err := o.getObjectGraph(namespace)
```
Hmmmm very good point. If there are existing objects in that namespace, it will fail since it will attempt to restore any objects that it finds in that object graph (including those that are already existing for that namespace). I think your suggestion to build from an empty object graph is a great idea.

fabriziopandini commented:
@jpmcb amazing work!
This is ready to go for me
Please squash commits

I will ask for more people to take a look at this at today office hours

jpmcb commented:
> Please squash commits

Done! Thank you for all the time spent on this!

mrajashree commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -592,6 +778,134 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
+}
+
+// Recreate all the OwnerReferences using the newUID of the owner nodes.
+func (o *objectMover) buildOwnerChain(obj *unstructured.Unstructured, n *node) {
```
does this also handle hierarchical owners? (obj A.ownerRef = obj B and obj B.ownerRef = obj C)

jpmcb replied:
I can't speak super well to the functionality of what this did before I refactored it into it's own function, but yes it should. This was simply yanked out of `createTargetObject` since it was duplicate code between that function, `restoreTargetObject`, and `backupTargetObject`. This can be found here: https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/client/cluster/mover.go#L531-L550

Once obj A is processed, it should have it's owner ref as obj B. Once obj B is processed, then it should then have it's owner obj as C. So, one by one, it should rebuild the entire owner tree.

mrajashree reviewed (commented):
this is great! just a few comments/questions

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +68,133 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
 	err := objectGraph.getDiscoveryTypes()
 	if err != nil {
-		return errors.Wrap(err, "failed to retrieve discovery types")
+		return nil, err
```
Good catch!! :+1:

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +344,70 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	// Save all objects group by group
+	log.Info("Saving files to %s", directory)
+	for groupIndex := 0; groupIndex < len(moveSequence.groups); groupIndex++ {
+		if err := o.backupGroup(moveSequence.getGroup(groupIndex), directory); err != nil {
```
Hmmm great question. 

I'm wondering if a partial restore of objects would leave a cluster in a poor state? For example, if first a `Cluster` failed to deploy, the next group in the move sequence would be `MachineDeployments`, then `MachineSets`. Shouldn't we fail early since objects in the move sequence are being moved in a per-defined order? Since we can easily attempt a restore maneuver again and objects that already exist (aka have already been restored) are skipped, I think it's reasonable to instruct users to simply re-run restore.

If the behavior was to continue to apply objects in the move group regardless of dependent objects being there, then I think we'd get into a really weird state where owner objects do not exist and they fail to reconcile.

mrajashree replied:
that makes sense, thanks

jpmcb commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +344,70 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+	}
+
+	// Reset the pause field on the Cluster object in the target management cluster, so the controllers start reconciling it.
+	log.V(1).Info("Resuming the target cluster")
```
Clusters that are being restored from files are _already in the pause state_.

- During a backup process, we set the cluster pause state to be true.
- Then we backup those objects to file (and the pause state will be reflected in the object yaml).
- We un-pause the clusters after a backup maneuver so they may be usable again for the current cluster.

At a latter time, we may need to restore those clusters again. Since those objects only exist as in-memory representations from the yaml files, we cannot call `setClusterPause` since there is no `Cluster` to pause in the target management cluster.

sbueringer commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +68,133 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+		return nil, err
+	}
+
+	var rawYamls [][]byte
```
```suggestion
	var rawYAMLs [][]byte
```
nit

sbueringer replied:
Also `processedYamls` below

sbueringer commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -60,46 +68,133 @@ func (o *objectMover) Move(namespace string, toCluster Client, dryRun bool) erro
+	return o.backup(objectGraph, directory)
+}
+
+func (o *objectMover) Restore(namespace string, toCluster Client, directory string) error {
```
namespace is unused

sbueringer commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -249,6 +344,70 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy) error {
+		}
+	}
+
+	// Reset the pause field on the Cluster object in the target management cluster, so the controllers start reconciling it.
```
I think that comment shouldbe adjusted a bit

sbueringer commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -493,6 +652,52 @@ func (o *objectMover) createGroup(group moveGroup, toProxy Proxy) error {
+	for i := range group {
+		nodeToCreate := group[i]
+
+		// Creates the Kubernetes object corresponding to the nodeToCreate.
```
```suggestion
		// Gets the Kubernetes object corresponding to the nodeToCreate.
```

sbueringer commented on cmd/clusterctl/client/cluster/mover.go:
```diff
@@ -493,6 +652,52 @@ func (o *objectMover) createGroup(group moveGroup, toProxy Proxy) error {
+func (o *objectMover) backupGroup(group moveGroup, directory string) error {
+	createTargetObjectBackoff := newWriteBackoff()
+	errList := []error{}
+	for i := range group {
+		nodeToCreate := group[i]
```
```suggestion
	for _, nodeToGet := range group {
```
similar in `restoreGroup`

jpmcb replied:
Good catch - ended up doing a bit more refactoring here. Also did the suggested code optimization for `createGroup`. Also renamed the object backoffs so they make more sense in the scope of `backup` and `restore`

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,455 @@ var moveTests = []struct {
+			graph := getObjectGraphWithObjs(tt.fields.objs)
+
+			// Get all the types to be considered for discovery
+			err := getFakeDiscoveryTypes(graph)
+			g.Expect(err).NotTo(HaveOccurred())
```
```suggestion
			g.Expect(getFakeDiscoveryTypes(graph)).To(Succeed())
```
nit

sbueringer replied:
(same in restore test)

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,455 @@ var moveTests = []struct {
+			// Write go string slice to directory
+			for _, file := range tt.files {
+				tempFile, _ := ioutil.TempFile(dir, "obj")
+				tempFile.Write([]byte(file))
+				tempFile.Close()
```
Please either explicitly handle or ignore the error (after a current PR from me is merged this will be a linter finding)

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,455 @@ var moveTests = []struct {
+			g.Expect(err).NotTo(HaveOccurred())
+
+			for _, obj := range objs {
+				graph.addRestoredObj(&obj)
```
Please handle the error

sbueringer commented on cmd/clusterctl/client/cluster/objectgraph.go:
```diff
@@ -110,6 +114,10 @@ func (n *node) isSoftOwnedBy(other *node) bool {
 }
 
+func (n *node) getFilename() string {
+	return n.identity.Kind + "_" + n.identity.Name + "_" + n.identity.Namespace + ".yaml"
```
I don't have a strong opinion on that but I would prefer `<kind>_<namespace>_<name>` (then it's "sorted" by namespace on a regular file listing)

jpmcb replied:
Sounds reasonable to me!

sbueringer commented on cmd/clusterctl/client/move.go:
```diff
@@ -38,6 +40,34 @@ type MoveOptions struct {
 	DryRun bool
 }
 
+// BackupOptions holds options supported by save
```
```suggestion
// BackupOptions holds options supported by backup
```

sbueringer commented on cmd/clusterctl/client/move.go:
```diff
@@ -38,6 +40,34 @@ type MoveOptions struct {
+
+	// Namespace where the objects describing the workload cluster exists. If unspecified, the current
+	// namespace will be used.
+	Namespace string
```
In case we really don't need it in the Restore func, we should drop it here

sbueringer commented on cmd/clusterctl/client/move_test.go:
```diff
@@ -94,6 +96,132 @@ func Test_clusterctlClient_Move(t *testing.T) {
+		client *fakeClient
+	}
+	// These tests are checking the Backup scaffolding
+	// The internal library handles the restore logic and tests can be found there
```
```suggestion
	// The internal library handles the backup logic and tests can be found there
```

sbueringer reviewed (commented):
@jpmcb Great work!. Mostly nits.

I think we should think about adding a e2e test for this. As we have already good coverage with unit tests I would suggest opening an issue for that. (@fabriziopandini WDYT?)

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,453 @@ var moveTests = []struct {
+			graph := getObjectGraphWithObjs(tt.fields.objs)
+
+			// Get all the types to be considered for discovery
+			g.Expect(getFakeDiscoveryTypes(graph)).NotTo(HaveOccurred())
```
```suggestion
			g.Expect(getFakeDiscoveryTypes(graph)).To(Succeed())
```
nit (reads a lot more fluent that way)

sbueringer replied:
(same in other cases below)

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,453 @@ var moveTests = []struct {
+			// Write go string slice to directory
+			for _, file := range tt.files {
+				tempFile, _ := ioutil.TempFile(dir, "obj")
+				tempFile.Write([]byte(file))
```
Maybe also handle that error

sbueringer commented on cmd/clusterctl/client/cluster/mover_test.go:
```diff
@@ -480,6 +486,453 @@ var moveTests = []struct {
+			// Write go string slice to directory
+			for _, file := range tt.files {
+				tempFile, _ := ioutil.TempFile(dir, "obj")
+				tempFile.Write([]byte(file))
```
same here

sbueringer reviewed (commented):
A few more nits, apart from that lgtm! (but let's also see what @fabriziopandini thinks about creating a follow-up issue for an e2e test)

jpmcb commented:
@sbueringer thank you for the thoughtful reviews!! Updated with your feedback :rocket:

sbueringer commented:
@jpmcb Looks good! Thank you for all the work! :)

/lgtm

fabriziopandini commented:
> let's also see what @fabriziopandini thinks about creating a follow-up issue for an e2e test

+1 😄 

/lgtm

jpmcb commented:
Awesome!! Very excited. Do we just need the approval label to get this in now?

vincepri reviewed (approved):
/approve

vincepri commented:
@jpmcb There is a couple of linting errors that need to be fixed before merging

jpmcb commented:
@vincepri Thanks for the heads up! Corrected the linting errors and pushed - let me know if this needs anything else!

fabriziopandini commented:
/lgtm
/approve
